Multiple templates per component, allowing arguments (#451)
ViewComponent/view_component: docs/guide/templates.md

  strip_trailing_whitespace(false)
end
```

## Sub-templates

Since 2.73.0
{: .label }

Experimental
{: .label .label-yellow }

ViewComponents can render sub-templates defined in the sidecar directory:

```text
app/components
├── ...
├── test_component.rb
├── test_component
|   ├── list.html.erb
|   └── summary.html.erb
├── ...
```

> Templates are compiled to methods in the format `render_#{template_basename}_template`, which can then be called in the component.

Reviewer: ```suggestion
Templates are compiled to private methods in the format `render_#{template_basename}_template`:
```

```ruby
class TestComponent < ViewComponent::Base
  def initialize(mode:)
    @mode = mode
  end

  def call
    case @mode
    when :list
      render_list_template
    when :summary
      render_summary_template
    end
  end
end
```

> To define which parameters a template accepts, use the
> [Rails Strict Locals](https://edgeguides.rubyonrails.org/action_view_overview.html#strict-locals) syntax.
> Unlike Rails, a missing `locals` declaration means the template takes no arguments.
> This option only affects sub-templates.

Reviewer: ```suggestion
To define which parameters a sub-template accepts, use the [Rails Strict Locals](https://edgeguides.rubyonrails.org/action_view_overview.html#strict-locals) syntax.

_Note: Unlike Rails, a missing `locals` declaration means the template takes no arguments._
```

```ruby
# test_component.rb
class TestComponent < ViewComponent::Base
  def initialize(mode:)
    @mode = mode
  end

  def call
    case @mode
    when :list
      render_list_template multiple: false
    when :multilist
      render_list_template multiple: true
    when :summary
      render_summary_template
    end
  end
end
```

```erb
<%# list.html.erb %>
<%# locals: (multiple:) %->
The parameter is <%= multiple %>

> ```

Reviewer: Might we put this example block above the Ruby? It's the more interesting bit IMO

  strip_trailing_whitespace(false)
end
```

> ## Multiple templates

Reviewer 2: What about "Partials"?

> ## Multiple templates

Reviewer 2: Or maybe "Sidecar partials" to remove ambiguity?

> ## Multiple templates

Reviewer: I like the term in theory, but in practice I think it's already "taken" by Rails' usage of it. As this feature is quite different from Rails partials, I'm hesitant to use the term.

Since 2.7.0
{: .label }

Experimental
{: .label .label-yellow }

ViewComponents can render multiple templates defined in the sidecar directory:

```text
app/components
├── ...
├── test_component.rb
├── test_component
|   ├── list.html.erb
|   └── summary.html.erb
├── ...
```

> Templates are compiled to methods in the format `render_template_#{template_basename}`, which can then be called in the component.

Author: It looks better indeed. It may make it a bit harder to grep for them but that shouldn't be a problem because they should be mostly private methods.
